### PR TITLE
wifi module

### PR DESCRIPTION
**Pull request by tjaartvdwalt**
This module displays some details on your wifi connection. It has 2 modes: ESSID and IP. You can toggle which one is displayed with a mouse click.

Toggling between the modes is very useful for saving space in the status bar, and is not possible (afaik) with the default i3status wifi module.

### CODE REVIEW

File: py3status/modules/wifi.py [hunk 1]
```diff
@@ -0,0 +1,84 @@
+"""
+Display the current wifi ESSID and IP address.
+It improves on the default i3status wifi module by allowing the user
```
Comment (ultrabug) [inline py3status/modules/wifi.py]: Add a blank line here

File: py3status/modules/wifi.py [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+import iwlib
+import socket
+from time import time
+
+
+class Py3status:
+    """
+    """
+    # available configuration parameters
+    cache_timeout = 30
```
Comment (ultrabug) [inline py3status/modules/wifi.py]: cache_timeout is not used in your code

File: py3status/modules/wifi.py [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+    mode = 'essid'
+    negative_cache_timeout = 2
```
Comment (ultrabug) [inline py3status/modules/wifi.py]: shouldn't it be cache_timeout instead to be in line with the other modules ? I don't see why the negative part

File: py3status/modules/wifi.py [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+        return response
+
+if __name__ == "__main__":
+    """
+    Test this module by calling it directly.
+    """
+    from time import sleep
+    x = Py3status()
+    config = {
+        'interface': 'wlan0',
```
Comment (ultrabug) [inline py3status/modules/wifi.py]: please, set it as a default with the other configuration parameters under the class definition instead

**Comment by ultrabug** [comment]
Interesting idea @tjaartvdwalt ! I have also the same need :) please see my comments

File: py3status/modules/wifi.py (hunk 3 above)
Comment (tjaartvdwalt) [inline py3status/modules/wifi.py]: This is sloppy code on my part. I copied the skeleton code from the `whatismyip` module, but I did not pay enough attention.

File: py3status/modules/wifi.py (hunk 2 above)
Comment (tjaartvdwalt) [inline py3status/modules/wifi.py]: I will fix this in my next commit. 
Is there any guide on what the default `cache_timeout` value should be?
Looking at other modules it seems to range anywhere from 5-30.

File: py3status/modules/wifi.py (hunk 1 above)
Comment (tjaartvdwalt) [inline py3status/modules/wifi.py]: Will do

File: py3status/modules/wifi.py (hunk 4 above)
Comment (tjaartvdwalt) [inline py3status/modules/wifi.py]: Will do

**Comment by pydsigner** [comment]
As far as cache_timeout defaults go, that depends greatly on how forgiveable stale information is versus how cheap it is to update. 10-15 seconds is quite fine here I think since getting the IP is cheap and unlikely to change often, though loss of wifi connectivity is something the user will want to know about sooner rather than later.

**Comment by ultrabug** [comment]
Ok, one thing about the code itself :
- maybe we need `format_ip` and `format_essid` parameters, where we have placeholders for users to decide what to show on the toggle ?

**Comment by ultrabug** [comment]
Other than that I gave some more thought about this because I always find it a pity to have to recode an i3status module because it lacks features...

Maybe the right answer to this PR and your feature @tjaartvdwalt is to actually develop a `toggle_group` or `scroll_group` or `module_group` whatever the name on py3status itself where users could **group multiple modules together** and switch between them using the mouse scroll buttons (4 and 5) !

This way users could still benefit from click events on mouse buttons 1, 2 and 3 on each module. My first idea would give something like this on i3status.conf for your use case (I need to use two different module instance _first_ and wlan0 to get them work on the same i3status output) :+1: 

````
wireless _first_ {
	format_down = "W"
	format_up = "%ip"
	scroll_group = 1
}

wireless wlan0 {
	format_down = "W"
	format_up = "%essid"
	scroll_group = 1
}
````

This would make py3status group the two outputs together and display one or the other on mouse scroll automatically.

What's interesting with this is that this would work with every module available and users could then group and *toggle* whatever modules they want together with a single line config.

If you and @pydsigner find this idea worth developing, I'll open a special RFC issue for it.

**Comment by tjaartvdwalt** [comment]
I agree that this is a better solution!

I will try to find some time over the weekend to look into the py3status code.

**Comment by ultrabug** [comment]
opened #178, closing here if that's ok. cheers

**Pull request closed without merging**